lib/core/src/UnliftIO/Compat.hs

```diff
@@ -61,3 +76,42 @@ handleIf
     -> m a
 handleIf f h = UnliftIO.handle
     (\e -> if f e then h e else UnliftIO.throwIO e)
+
+-- NOTE: This instance is problematic when parameter e is an Exception instance,
+-- and 'catchAny' or similar functions are used in the wrapper.
+--
+-- See: https://github.com/fpco/unliftio/issues/68
+instance (MonadUnliftIO m, Typeable e) => MonadUnliftIO (ExceptT e m) where
+  withRunInIO exceptToIO =
+    withExceptT unInternalException $ ExceptT $ try $
+      withRunInIO $ \runInIO ->
+        exceptToIO
+          (runInIO . (either (throwIO . InternalException) pure <=< runExceptT))
+
+-- | Wrapper for Left results of runExceptT. Used to deliver return values in
+-- the MonadUnliftIO instance of ExceptT.
+newtype InternalException e = InternalException { unInternalException :: e }
+    deriving Typeable
+instance Show (InternalException e) where
+    show _ = "MonadUnliftIO InternalException"
+instance Typeable e => Exception (InternalException e)
```

Comment:
I wonder if we aren't going to shoot ourselves in the foot with this... We catch `SomeException` in a few places and this instances introduces a quite confusing behavior... I wonder if we couldn't leverage the fact that:

a) `m` will always be `IO` for us. 
b) Async exception aren't caught by unliftio functions. 

to write it as:

```hs
instance (MonadUnliftIO m, MonadIO m, Typeable e, Show e) => MonadUnliftIO (ExceptT e m) where
    withRunInIO exceptToIO =
        unSmuggleExceptT $ ExceptT $
            withRunInIO $ \runInIO -> Control.Exception.try $
                exceptToIO (runInIO . smuggleExceptT . runExceptT)
      where
        smuggleExceptT :: m (Either e a) -> m a
        smuggleExceptT = (=<<) $ either
            (liftIO . Control.Exception.throwIO . SomeAsyncException . InternalException)
            pure

        unSmuggleExceptT :: ExceptT SomeAsyncException m a -> ExceptT e m a
        unSmuggleExceptT = mapExceptT $ \inner -> inner >>= \case
            Right a -> return $ Right a
            Left asyncE@(SomeAsyncException e) ->
                case toInternalException e of
                    Just (InternalException e) -> return (Left e)
                    Nothing -> liftIO $ Control.Exception.throwIO asyncE

        toInternalException :: Exception any => any -> Maybe (InternalException e)
        toInternalException =
            Control.Exception.fromException . Control.Exception.toException
```

This way, we properly handle the ExceptT checked exception and other exceptions:

```hs
-- | Left MyException
case1 :: ExceptT MyException IO String
case1 = handle handler $ do
    throwE MyException
  where
    handler (_e :: MyException) = return "Caught"

-- | Left MyException
case2 :: ExceptT MyException IO String
case2 = handleAny handler $ do
    throwE MyException
  where
    handler (_e :: SomeException) = return "Caught"

-- | Right "Caught"
case3 :: ExceptT MyException IO String
case3 = handle handler $ do
    liftIO $ throwIO MyException
  where
    handler (_e :: MyException) = return "Caught"

-- | Right "Caught"
case4 :: ExceptT MyException IO String
case4 = handleAny handler $ do
    liftIO $ throwIO MyException
  where
    handler (_e :: SomeException) = return "Caught"
```

Otherwise, `case2` would return a `Right "Caught"` as well which is misleading.

Reply:
I don't think this instance is safe at all or behave as we would expect. For instance:

```hs
main :: IO ()
    print =<< runExceptT case1
    print =<< runExceptT (do
        askUnliftIO >>= \u -> liftIO (unliftIO u case1)
        )
 ```
  
This prints:

```
Left MyException
Main.hs: UnliftIOControlException {unUnliftIOControlException = MyException}
```

Where I'd expect the same behavior. Though, stepping back a bit, what would it even mean to run an ExceptT in IO? It doesn't make sense in the end because there's no way that we can capture the context regarding the checked exception in IO only. So, what would it mean to run `throwE MyException` in IO? At best it means "throw the exception", whereas the semantic given by ExceptT would be more "return a Left". But we can interpret "return a Left" is mere IO because we can only express `IO a`. 

We would need a generalized definition of `MonadUnliftIO` to make it work properly, something like:


```hs
class Monad m => MonadUnliftIO m where
    type F m :: * -> *
    
    withRunInIO :: ((forall a. m a -> IO (F m a)) -> IO b) -> m b 
```

And then define:

```
instance MonadUnliftIO m => MonadUnliftIO (ReaderT r m a) where
   type F (ReaderT r m) = Identity
   
instance MonadUnliftIO m => MonadUnliftIO (ExceptT e m a) where
   type F (ExceptT r m) = Either e
```

Without that, I don't see how we can write a sound instance for `ExceptT`.

Reply:
Thanks for the analysis.
I will remove the `MonadUnliftIO` instances.
If we are going to define this instance, it should work generally.